hoomd/mpcd/pytest/test_stream.py

```diff
@@ -517,3 +523,216 @@ def test_check_mpcd_particles(self, simulation_factory, snap):
             geometry=hoomd.mpcd.geometry.PlanarPore(separation=8.0, length=7.0),
         )
         assert not ig.streaming_method.check_mpcd_particles()
+
+
+class TestSphere:
+
+    def _make_particles(self, snap):
+        if snap.communicator.rank == 0:
+            snap.mpcd.N = 4
+
+            # particle 1: Hits the wall in the second streaming step, gets
+            # reflected accordingly
+            snap.mpcd.position[0] = [2.85, 0.895, np.sqrt(6) + 0.075]
+            snap.mpcd.velocity[0] = [1., 0.7, -0.5]
+
+            # particle 2: Always inside the sphere, so no reflection by the BC
+            snap.mpcd.position[1] = [0., 0., 0.]
+            snap.mpcd.velocity[1] = [-1., -1., -1.]
+
+            # particle 3: Hits the wall normally and gets reflected back.
+            snap.mpcd.position[2] = 0.965 * np.array([-1., -2., np.sqrt(11)])
+            snap.mpcd.velocity[2] = 0.25 * np.array([-1., -2., np.sqrt(11)])
+
+            # particle 4: Lands almost exactly on the sphere surface and needs
+            # to be backtracked one complete step
+            snap.mpcd.position[3] = [1.92, -1.96, -np.sqrt(8) + 0.05]
+            snap.mpcd.velocity[3] = [0.8, -0.4, -0.5]
+        return snap
+
+    def test_step_noslip(self, simulation_factory, snap):
+        """Test step with no-slip boundary conditions."""
+        snap = self._make_particles(snap)
+        sim = simulation_factory(snap)
+        sm = hoomd.mpcd.stream.BounceBack(
+            period=1, geometry=hoomd.mpcd.geometry.Sphere(radius=4.0))
+        ig = hoomd.mpcd.Integrator(dt=0.1, streaming_method=sm)
+        sim.operations.integrator = ig
+
+        sim.run(1)
+        snap = sim.state.get_snapshot()
+        if snap.communicator.rank == 0:
+            np.testing.assert_array_almost_equal(
+                snap.mpcd.position[0],
+                [2.95, 0.965, np.sqrt(6) + 0.025])
+            np.testing.assert_array_almost_equal(snap.mpcd.velocity[0],
+                                                 [1., 0.7, -0.5])
+            np.testing.assert_array_almost_equal(snap.mpcd.position[1],
+                                                 [-0.1, -0.1, -0.1])
+            np.testing.assert_array_almost_equal(snap.mpcd.velocity[1],
+                                                 [-1., -1., -1.])
+            np.testing.assert_array_almost_equal(
+                snap.mpcd.position[2],
+                0.99 * np.array([-1., -2., np.sqrt(11)]))
+            np.testing.assert_array_almost_equal(
+                snap.mpcd.velocity[2],
+                0.25 * np.array([-1., -2., np.sqrt(11)]))
+            np.testing.assert_array_almost_equal(snap.mpcd.position[3],
+                                                 [2., -2., -np.sqrt(8)])
+            np.testing.assert_array_almost_equal(snap.mpcd.velocity[3],
+                                                 [0.8, -0.4, -0.5])
```

Comment:
Can this set of checks be placed in its own function? It appears to be checking the same things many times throughout different tests, but with different values each time.

It may even be useful to do the repeated checks in a loop.

Reply:
I think that `np.testing.assert_array_almost_equal` is more or less that function, since I would need to pass lists of positions and velocities. I can transpose the checks so that there are only two calls to `np.testing.assert_array_almost_equal` if you want to reduce the number of those calls, but this pattern of checking each particle individually is used in other spots in this test file. I think the original reason for writing it this way was that it was easier to figure out which particle was causing the error, likely when it was using `unittest` rather than `pytest`.

Reply:
I rearranged these units to reduce the number of assertions, so this should be addressed. For debugging these geometries, having explicit lists of coordinates that can be referred to on each step is helpful, so I don't think any further abstraction should be done.